### PR TITLE
React Hooks v16.8

### PR DESCRIPTION
**Pull request by Skolotsky**
Wrappers for react hooks and way to create functional component like this
```kotlin
val renderCount = functionalComponent<RProps> {
    val (count, setCount) = useState(0)
    useEffect {
        console.log("useEffect $count")
    }
    +"count $count"
    button {
        attrs {
            onClickFunction = {
                setCount(count + 1)
            }
        }
    }
}

fun RBuilder.renderCount() {
    child(renderCount)
}
```

### CODE REVIEW

File: kotlin-react/src/main/kotlin/react/Imports.kt [hunk 1]
```diff
@@ -86,3 +86,49 @@ external fun <P : RProps> lazy(loadComponent: () -> Promise<RClass<P>>): RClass<
...
+
+// State Hook
+@JsName("useState")
+external fun <T> rawUseState(initValue: T): RDependenciesArray
+
+// Reducer Hook
+@JsName("useReducer")
+external fun <S, A> rawUseReducer(reducer: RReducer<S, A>, initialState: S): RDependenciesArray
+@JsName("useReducer")
+external fun <S, A> rawUseReducer(reducer: RReducer<S, A>, initialState: S, initialAction: A): RDependenciesArray
```
Comment (ezhaka) [inline kotlin-react/src/main/kotlin/react/Imports.kt]: Can we use `initialAction: A = definedExternally` and avoid duplicating the function?

Comment (Skolotsky) [reply kotlin-react/src/main/kotlin/react/Imports.kt]: I don't sure that it's equal in converted js. And this "optimisation" is unnecessary.

**Comment by Hypnosphi** [comment]
Hooks should only be defined inside a functional component body, so it's better to define them as extension functions

File: kotlin-react/src/main/kotlin/react/RBuilder.kt [hunk 2]
```diff
@@ -120,3 +120,29 @@ fun <P : RProps> forwardRef(handler: RBuilder.(RProps, RRef) -> Unit): RClass<P>
         buildElements { handler(props, ref) }
     }
 }
+
+typealias FunctionalComponent<P> = (props: P) -> dynamic
+
+/**
+ * Get functional component from [func]
+ */
+fun <P : RProps> functionalComponent(
```
Comment (Hypnosphi) [inline kotlin-react/src/main/kotlin/react/RBuilder.kt]: Does it replace `rFunction`?
https://github.com/JetBrains/kotlin-wrappers/blob/master/kotlin-react/src/main/kotlin/react/ReactComponent.kt#L78-L85

**Pull request merged**